webrender/src/clip_scroll_node.rs

```diff
@@ -436,40 +457,71 @@ impl ClipScrollNode {
         viewport_scroll_offset: &LayerVector2D,
         viewport_rect: &LayerRect,
     ) -> LayerVector2D {
-        let sticky_frame_info = match self.node_type {
-            NodeType::StickyFrame(info, _) => info,
+        let info = match self.node_type {
+            NodeType::StickyFrame(ref info) => info,
             _ => return LayerVector2D::zero(),
         };
 
+        if info.margins.top.is_none() && info.margins.bottom.is_none() &&
+            info.margins.left.is_none() && info.margins.right.is_none() {
+            return LayerVector2D::zero();
+        }
+
+        // The viewport and margins of the item establishes the maximum amount that it can
+        // be offset in order to keep it on screen. Since we care about the relationship
+        // between the scrolled content and unscrolled viewport we adjust the viewport's
+        // position by the scroll offset in order to work with their relative positions on the
+        // page.
         let sticky_rect = self.local_viewport_rect.translate(viewport_scroll_offset);
-        let mut sticky_offset = LayerVector2D::zero();
 
-        if let Some(info) = sticky_frame_info.top {
-            sticky_offset.y = viewport_rect.min_y() + info.margin - sticky_rect.min_y();
-            sticky_offset.y = sticky_offset.y.max(0.0).min(info.max_offset);
+        let mut sticky_offset = LayerVector2D::zero();
+        if let Some(margin) = info.margins.top {
+            // First we calculate the potential offset if we are scrolled against the top edge
+            // of the viewport. This would be enough to move us to the edge minus the margin.
+            // Recall that we are working with the position of the viewport (which is not scrolled)
+            // relative the scrolled sticky rect.
+            sticky_offset.y = viewport_rect.min_y() + margin - sticky_rect.min_y();
+
+            // If the offset is less than zero, that means it is already below the top of the
+            // viewport rectangle.
+            sticky_offset.y = sticky_offset.y.max(0.0);
```

Comment:
This code is much clearer than it was before, thanks! It makes sense to me, although as I was reading it it occurred to me it sometimes computes a thing for sticky_offset, but then cancels the thing it just computed in the max call. It's effectively equivalent to this:

```
// The desired position of the item assuming we are scrolled against the top edge of the viewport
let desired_position = viewport_rect.min_y() + margin;
// If sticky rect is closer to the edge than desired, adjust the offset to put it at the desired location
if sticky_rect.min_y() < desired_position {
    sticky_offset.y = desired_position - sticky_rect.min_y();
}
```
I'm not sure if they compile down to the same thing but avoiding the setting and re-setting of sticky_offset.y might be good.

Reply:
Sure. This seems like a reasonable change. I think I was just trying to avoid another level of indentation here.